### PR TITLE
test/checks: improve AWS connection platform checks

### LINKED ISSUE #23055: Support `AWS` connections
## Context

We plan to build an AWS S3 sink, in which we write to an S3 bucket under the end user's control: https://github.com/MaterializeInc/materialize/issues/7256.

We'll need to integrate with AWS IAM so that our users can grant their Materialize region the ability to write to their desired S3 bucket.

Integration with AWS IAM would also be useful for facilitating IAM authentication with the following existing source types:

  * PostgreSQL on AWS RDS (https://github.com/MaterializeInc/materialize/issues/23056)
  * Kafka on AWS MSK (https://github.com/MaterializeInc/materialize/issues/8237)

While we support AWS RDS/MSK today via traditional username/password authentication, many users prefer to use AWS IAM with these services, as it eliminates the need to manage additional user accounts/passwords.

## Design sketch

See the [AWS connector](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20220404_connectors.md#aws-connector) in the original `CREATE CONNECTION` design document for prior art.

Here's a version that's updated for what we've learned since that design document was written 1.5 years ago:

```sql
CREATE CONNECTION <connector_name> TO AWS (
  ACCESS KEY ID [[=] <value>],
  SECRET ACCESS KEY [[=] SECRET <value>],
  TOKEN [[=] SECRET <value>],
  ASSUME ROLE ARN [[=] <value>],
  ASSUME ROLE SESSION NAME [[=] <value>]
);
```

There are two ways to use this connection:

  1. You can specify `ACCESS KEY ID`, `SECRET ACCESS KEY`, and possibly `TOKEN`. This way is a bit sketchy, as you have  to punch your access credentials into Materialize, but it's secure enough via our secrets support. After all, punching in usernames/passwords is exactly what you have to do to connect to PostgreSQL/Kafka today.
  2. You can specify `ASSUME ROLE ARN`, and optionally `ASSUME ROLE SESSION NAME`, and use IAM policies in your target AWS account to allow Materialize to assume that role in your account.

Option 2 is the preferred way for production/sensitive connections. Option 1 is easier to get up and running with in development for nonsensitive connections (e.g., to a test/staging account).

Here's an example of using option 2:

1. You create a role in your target AWS account with a name of your choosing, e.g. `MaterializeS3`. Note down its ARN, e.g., `arn:aws:iam::001234567890:role/MaterializeS3Conn`.

2. You create the connection in Materialize:
    
    ```sql
    CREATE CONNECTION s3_conn TO AWS (
        ASSUME ROLE ARN = 'arn:aws:iam::001234567890:role/MaterializeS3Conn'
    );
    ```
    
    Note down the connection ID, e.g., u45.

3. You add a trust policy to the role that allows your Materialize deployment to access that role:

    ```json
    {
      "Version": "2012-10-17",
      "Statement": [
        {
          "Sid": "",
          "Effect": "Allow",
          "Principal": {
            "AWS": "664411391173"
          },
          "Action": "sts:AssumeRole",
          "Condition": {
            "StringEquals": {
              "sts:ExternalId": "6f11558e-89da-4733-9a24-36eb9fd58129_u45"
            }
          }
        }
      ]
    }
    ```

4. Use the AWS connection as you like, e.g., to copy to S3: `COPY ... TO 's3://...' USING AWS CONNECTION s3_conn`. Materialize assumes the `MaterializeS3Conn` role in your account, as directed, and uses that role to write the data to your S3 bucket. 

Under the hood, Materialize will use the AWS credentials provided to create an AWS SDK client for the appropriate service. We have old code for this from when we supported S3 sources that we can dredge up. It's pretty straightforward: you plug the access key or assume role configuration into the SDK pretty directly.

One part that requires close attention to detail is external ID handling. When assuming a role, to make things safe, Materialize needs to specify a user ID that identifies from which Materialize region the request originates, as well as which _connection_ within that organization. The cloud already passes in the AWS external ID _prefix_ to use for the region (always a randomly generated UUID, [as AWS recommends](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html)). But Materialize needs to to be sure to append the connection ID to the external ID. This allows end users to write trust policies that ensure a given role can be assumed by only the _one_ identified connection in their Materialize account—and then they can use RBAC in Materialize to control who has `USAGE` privileges on that connection. Described in much more detail here: https://github.com/MaterializeInc/cloud/issues/2837

The upshot is that the external ID Materialize should use with assume role-style AWS connections is:

```
<external-id-prefix-provided-by-cloud>_<connection-id>
```

## See also

* The equivalent in Snowflake is [`CREATE STORAGE INTEGRATION`](https://docs.snowflake.com/en/sql-reference/sql/create-storage-integration) and [Configuring Secure Access to Amazon S3](https://docs.snowflake.com/en/user-guide/data-load-s3-config)
* https://github.com/MaterializeInc/cloud/issues/2837
* https://github.com/MaterializeInc/cloud/issues/7920
```[tasklist]
### Work Items
- [ ] https://github.com/MaterializeInc/materialize/pull/23173
- [x] Put AWS Connection retroactively behind feature flag
- [ ] https://github.com/MaterializeInc/materialize/pull/23626
- [ ] https://github.com/MaterializeInc/materialize/pull/23282
- [ ] https://github.com/MaterializeInc/cloud/issues/8191
- [ ] https://github.com/MaterializeInc/materialize/pull/23957
- [ ] https://github.com/MaterializeInc/materialize/pull/23958
- [ ] #23963
```

### PR DESCRIPTION
**Pull request by benesch**
Various improvements to AWS connection testing:

  * Add end-to-end tests that run against real AWS for both credentials and assume role authentication, with both positive (expected to succeed) and negative (expected to fail) test cases.
  * Structure the validation error when a role fails to constrain the external ID, in order to give users more guidance on what's wrong and how to fix it.

### Motivation

Fixes #23055.

### Tips for reviewer

Go commit by commit. Details in commit messages.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

**Comment by benesch** [comment]
@sploiselle @petrosagg @rjobanp — if y'all are working this week, can I ask one of you to take a look at this PR? Hoping to get it into this week's release so that we can put a bow on all the AWS connection work this year, teeing up the COPY TO/FROM S3 work for next year. 

Don't be scared off by the diffstat; it's almost entirely code movement and lockfile updates. Definitely go commit by commit. The only things that need close review are:

  1. The penultimate commit's addition of test/aws/mzcompose.py, which is a new E2E test for AWS connections. Everything else in the commit is `cargo vet` noise.
  2. The entirety of the final commit, which improves the `VALIDATE CONNECTION` error message.

**Comment by benesch** [comment]
Nightlies were successful on the new AWS test: https://buildkite.com/materialize/nightlies/builds/5715

**Review by sploiselle — APPROVED** [review]
- nit isn't worth going back through CI
- Should we close out some of the TODOs in `aws-connection.td`? e.g. https://github.com/MaterializeInc/materialize/blob/main/test/aws/aws-connection/aws-connection.td#L65
- I don't have any special knowledge of the jump role, external ID scheme for AWS permission control. the tests are convincing but I haven't followed this work closely enough to know what to expect.

Approving but consider more of a +1 on the AWS integration test

File: src/storage-types/src/connections/aws.rs [hunk 1]
```diff
@@ -13,9 +13,12 @@ use anyhow::{anyhow, bail};
 use aws_config::sts::AssumeRoleProvider;
 use aws_credential_types::provider::{ProvideCredentials, SharedCredentialsProvider};
 use aws_credential_types::Credentials;
+use aws_sdk_sts::error::SdkError;
+use aws_sdk_sts::operation::get_caller_identity::GetCallerIdentityError;
 use aws_types::region::Region;
 use aws_types::SdkConfig;
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/storage-types/src/connections/aws.rs:20]: Good eyes, thanks!

File: src/storage-types/src/connections/aws.rs [hunk 2]
```diff
@@ -386,3 +380,36 @@ impl AwsConnection {
...
+            } => Some(format!("The trust policy for the connection's role ({role_arn}) is insecure and allows any Materialize customer to assume the role.")),
+            _ => None
+        }
+    }
+
+    /// Reports a hint for the user about how the error could be fixed.
+    pub fn hint(&self) -> Option<String> {
+        match self {
+            AwsConnectionValidationError::RoleDoesNotRequireExternalId { .. } => {
+                Some("See: https://materialize.com/s/aws-connection-role-trust-policy".into())
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/storage-types/src/connections/aws.rs:410]: Yeah it's a shortlink added in this PR, so as long as I didn't typo (🤞🏽) it'll start working once I merge this.

**Comment by benesch** [comment]
> * nit isn't worth going back through CI

Cleaned it up nonetheless!

> * Should we close out some of the TODOs in `aws-connection.td`? e.g. https://github.com/MaterializeInc/materialize/blob/main/test/aws/aws-connection/aws-connection.td#L65

Oh good spot, thanks. Replaced those TODOs with a pointer to /test/aws/.

> * I don't have any special knowledge of the jump role, external ID scheme for AWS permission control. the tests are convincing but I haven't followed this work closely enough to know what to expect.
> 
> Approving but consider more of a +1 on the AWS integration test

Ack. @moulimukherjee @guswynn — would appreciate a more in-depth post-merge look at this PR when you're back. Going to merge on @sploiselle's approval, though, since this is adding tests for a feature that's already merged so this strictly improves the situation.

File: test/aws/mzcompose.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        )
+
+        for fn in [test_credentials, test_assume_role]:
+            with c.test_case(fn.__name__):
+                fn(c, ctx)
+
+
+def test_credentials(c: Composition, ctx: TestContext):
+    # Create a user with an access key.
+    customer_user = f"testdrive-{ctx.seed}-Customer"
```
Comment (moulimukherjee) [inline test/aws/mzcompose.py:126]: Should the users and roles created in this test be deleted as well as a cleanup after the test? If I am understanding right, these are actually created in Materialize Scratch account whenever they are run at least locally and we'll someday hit the limit?

image

**Pull request merged**